### ISSUE TITLE
FTP (Alternative) path fails with spaces and other special characters

### ISSUE THREAD

**Issue by ts678** [duplicati/duplicati#4587]
- [x] I have searched open and closed issues for duplicates.
- [x] I have searched the [forum](https://forum.duplicati.com) for related topics.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.5.114 Canary but this is a very old bug
- **Operating system**: Windows 10 Version 21H1
- **Backend**: FTP (Alternative)

## Description

The typical nuisance is that a path with embedded spaces (e.g. matching the job name) fails because a space gets percent-encoded as %20 when Duplicati tries to access the server. I looked further today, and I think it's a generalized accident of URI encoding methods.

## Steps to reproduce

1. Set path to space= delims=leftanglerightanglehash%doublequoteunwise={}verticalbarbackslash^[]`
2. Make a folder with this name. I'm on Windows which unfortunately has more forbidden characters.
3. On Destination screen, try `Test connection` button.

- **Actual result**:
Failed to connect: CWD failed. "/space=%20delims=leftanglerightanglehash%25doublequoteunwise=%7B%7Dverticalbarbackslash%5E[]%60": directory not found.
- **Expected result**:
Connection worked! (which is what the Test connection popup says if I change Storage Type to FTP)

Wireshark for FTP shows:

CWD space= delims=leftanglerightanglehash%doublequoteunwise={}verticalbarbackslash^[]`
250 CWD successful. "/space= delims=leftanglerightanglehash%doublequoteunwise={}verticalbarbackslash^[]`" is current directory.

Wireshark for FTP (Alternative) shows:

CWD /space=%20delims=leftanglerightanglehash%25doublequoteunwise=%7B%7Dverticalbarbackslash%5E[]%60
550 CWD failed. "/space=%20delims=leftanglerightanglehash%25doublequoteunwise=%7B%7Dverticalbarbackslash%5E[]%60": directory not found.

A debugger breakpoint after the code

https://github.com/duplicati/duplicati/blob/265d202a1f99f12f7c5bcf8b9e0deac50ac799c3/Duplicati/Library/Backend/AlternativeFTP/AlternativeFTPBackend.cs#L532-L534

```
this._url "ftp://localhost/space= delims=leftanglerightanglehash%doublequoteunwise={}verticalbarbackslash^[]`/"
uri {ftp://localhost/space= delims=leftanglerightanglehash%25doublequoteunwise={}verticalbarbackslash^[]`/}
uri.AbsolutePath "/space=%20delims=leftanglerightanglehash%25doublequoteunwise=%7B%7Dverticalbarbackslash%5E[]%60/"
```

[Uri Class](https://docs.microsoft.com/en-us/dotnet/api/system.uri?view=netframework-4.8)

> By default, any reserved characters in the URI are escaped in accordance with RFC 2396.

[Excluded US-ASCII Characters](https://datatracker.ietf.org/doc/html/rfc2396#section-2.4.3) is where I got `space delims unwise`, but I couldn't put them in Windows to see `FTP` handle them.
The # character had to go because URI parsing pulls it out to be a fragment identifier. Debugger information from an earlier test is:

```
uri.AbsoluteUri "ftp://localhost/space=%20delims=%3C%3E#%25%22unwise=%7B%7D%7C%5C%5E[]%60/"
uri.AbsolutePath "/space=%20delims=%3C%3E"
uri.Fragment "#%25%22unwise=%7B%7D%7C%5C%5E[]%60/"
```

Issue fix may require staying away from dangerous System.Uri. I see that code currently sometimes uses a Duplicati-written parser:

https://github.com/duplicati/duplicati/blob/265d202a1f99f12f7c5bcf8b9e0deac50ac799c3/Duplicati/Library/Utility/Uri.cs#L32-L36

The counter-argument is that FluentFTP apparently knows how to handle a URI, at least in some cases, such as the following code:

https://github.com/robinrodricks/FluentFTP/blob/1583fe6d1992ca99a30687643f8991302b639d45/FluentFTP/Client/FtpClient_Static.cs#L5-L14

so I'll just leave it as a comment that the path corruption looks like something got URL percent-encoded that should not have been.

**Comment by warwickmm** [comment]
The [Uri.GetComponents](https://docs.microsoft.com/en-us/dotnet/api/system.uri.getcomponents?view=netframework-4.7.1) method might be useful.  That should allow us to extract the unescaped Host, Port, AbsoluteUri, etc.

I don't have an FTP server to test with, but might be able to set one up if I have time.

**Comment by warwickmm** [comment]
> Actual result:
Failed to connect: CWD failed. "/space=%20delims=leftanglerightanglehash%25doublequoteunwise=%7B%7Dverticalbarbackslash%5E[]%60": directory not found.

I can reproduce this, but with a slightly different error message: `Failed to connect: Failed to change directory.`  It looks like the error message could be server-dependent:

https://github.com/robinrodricks/FluentFTP/blob/01131bbd99a3c4dbb0f4dbf0281c4784c9eb3406/FluentFTP/Client/FtpClient_FolderManagement.cs#L795-L796

**Comment by warwickmm** [comment]
Issue #1885 could be related.

**Comment by warwickmm** [comment]
The changes in pull request #4588 allow me to use `space\=\ delims\=leftanglerightanglehash%doublequoteunwise\={}verticalbarbackslash^[]` as the destination folder.

**Comment by ts678** [comment]
>  It looks like the error message could be server-dependent:

and it looks like Duplicati just passes along the server's message, or at least my Wireshark error and Duplicati's look the same...

> Issue #1885 could be related.

Not impossible, however that was `FTP` (which works now, and its GitHub history shows no obvious relevant changes), not `FTP (Alternative)`. I also had GitHub compare v2.0.1.17-2.0.1.17_canary_2016-05-14 and v2.0.1.18-2.0.1.18_canary_2016-08-10. Looking at the file names changed still didn't explain to me the very clear statement in the issue that it broke at exactly 2.0.1.18.

> The changes in pull request #4588 allow me to use

Wonderfully fast work. Thanks so much! Just getting the space healthy would save me from having to workaround the file paths.
Other special characters tend not to be used in my work, however for completeness, I'll make some further comments on them...
 
Note that you *could* test other special characters because Linux FS is more tolerant. If not, I might try Wireshark on next Canary.
Even if the server can't find things (because of a request for impossible characters), I can still see whether request was done right.
On Windows, I'm actually using an FTP server (FileZilla Server) that shows me the request. I don't know if yours is able to do that.

Hopefully the PR fixes at least all ASCII printables. FluentFTP does appear to do [UTF8](https://datatracker.ietf.org/doc/html/rfc2640), but server feature support is required for it.

**Comment by warwickmm** [comment]
I ran a few more tests and the changes allow me to use `space= delims=<>%"unwise={}|^[]` as the folder.  I wasn't able to spend  much time debugging, but it appears that `Uri.AbsolutePath` strips things after (and including) `#`, and changes `\` to `/`.

**Comment by ts678** [comment]
> strips things after (and including) #

That's removal of the fragment identifier, mentioned in the original post (though in a different context), and documented [here](https://docs.microsoft.com/en-us/dotnet/api/system.uri?view=netframework-4.8#remarks).

> changes \ to /

I think I saw that too (in a different context), but because Windows won't allow the / I had dodged that by not using backslash.
So I don't think it's a change from previous behavior, and even if the # character is still broken, there are many fixed characters.
